File: packages/ts/src/tx/codecs.ts
/**
 * Hand-written protobuf codecs for the `qorechain.rdk.v1` transaction messages.
 *
 * These mirror the on-chain `rdk` module's `tx.proto` exactly — field numbers,
 * wire types, and message names — and follow the `cosmjs-types` generated-code
 * shape (`encode` / `decode` / `fromPartial`) so they register directly with a
 * `@cosmjs` `Registry`. 64-bit integers are represented as `bigint`, and `bytes`
 * fields as `Uint8Array`.
 */
import { BinaryReader, BinaryWriter } from "cosmjs-types/binary";

const TYPE_URL_PREFIX = "/qorechain.rdk.v1.";

export interface MsgCreateRollup {
  creator: string;
  rollupId: string;
  profile: string;
  vmType: string;
  stakeAmount: bigint;
}
export interface MsgCreateRollupResponse {
  rollupId: string;
}
export interface MsgSubmitBatch {
  sequencer: string;
  rollupId: string;
  batchIndex: bigint;
  stateRoot: Uint8Array;
  prevStateRoot: Uint8Array;
  txCount: bigint;
  dataHash: Uint8Array;
  proof: Uint8Array;
  withdrawalsRoot: Uint8Array;
}
export interface MsgChallengeBatch {
  challenger: string;
  rollupId: string;
  batchIndex: bigint;
  proof: Uint8Array;
}
export interface MsgResolveChallenge {
  resolver: string;
  rollupId: string;
  batchIndex: bigint;
  fraudUpheld: boolean;
}
export interface MsgPauseRollup {
  creator: string;
  rollupId: string;
  reason: string;
}
export interface MsgResumeRollup {
  creator: string;
  rollupId: string;
}
export interface MsgStopRollup {
  creator: string;
  rollupId: string;
}
export interface MsgExecuteWithdrawal {
  submitter: string;
  rollupId: string;
  batchIndex: bigint;
  withdrawalIndex: bigint;
  recipient: string;
  denom: string;
  amount: bigint;
  proof: Uint8Array[];
}

export const MsgCreateRollup = {
  typeUrl: `${TYPE_URL_PREFIX}MsgCreateRollup`,
  encode(message: MsgCreateRollup, writer: BinaryWriter = BinaryWriter.create()): BinaryWriter {
    if (message.creator !== "") writer.uint32(10).string(message.creator);
    if (message.rollupId !== "") writer.uint32(18).string(message.rollupId);
    if (message.profile !== "") writer.uint32(26).string(message.profile);
    if (message.vmType !== "") writer.uint32(34).string(message.vmType);
    if (message.stakeAmount !== 0n) writer.uint32(40).int64(message.stakeAmount);
    return writer;
  },
  decode(input: BinaryReader | Uint8Array, length?: number): MsgCreateRollup {
    const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
    const end = length === undefined ? reader.len : reader.pos + length;
    const message: MsgCreateRollup = {
      creator: "",
      rollupId: "",
      profile: "",
      vmType: "",
      stakeAmount: 0n,
    };
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.creator = reader.string();
          break;
        case 2:
          message.rollupId = reader.string();
          break;
        case 3:
          message.profile = reader.string();
          break;
        case 4:
          message.vmType = reader.string();
          break;
        case 5:
          message.stakeAmount = reader.int64();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object: Partial<MsgCreateRollup>): MsgCreateRollup {
    return {
      creator: object.creator ?? "",
      rollupId: object.rollupId ?? "",
      profile: object.profile ?? "",
      vmType: object.vmType ?? "",
      stakeAmount: object.stakeAmount ?? 0n,
    };
  },
};

export const MsgCreateRollupResponse = {
  typeUrl: `${TYPE_URL_PREFIX}MsgCreateRollupResponse`,
  encode(
    message: MsgCreateRollupResponse,
    writer: BinaryWriter = BinaryWriter.create(),
  ): BinaryWriter {
    if (message.rollupId !== "") writer.uint32(10).string(message.rollupId);
    return writer;
  },
  decode(input: BinaryReader | Uint8Array, length?: number): MsgCreateRollupResponse {
    const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
    const end = length === undefined ? reader.len : reader.pos + length;
    const message: MsgCreateRollupResponse = { rollupId: "" };
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.rollupId = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object: Partial<MsgCreateRollupResponse>): MsgCreateRollupResponse {
    return { rollupId: object.rollupId ?? "" };
  },
};

export const MsgSubmitBatch = {
  typeUrl: `${TYPE_URL_PREFIX}MsgSubmitBatch`,
  encode(message: MsgSubmitBatch, writer: BinaryWriter = BinaryWriter.create()): BinaryWriter {
    if (message.sequencer !== "") writer.uint32(10).string(message.sequencer);
    if (message.rollupId !== "") writer.uint32(18).string(message.rollupId);
    if (message.batchIndex !== 0n) writer.uint32(24).uint64(message.batchIndex);
    if (message.stateRoot.length !== 0) writer.uint32(34).bytes(message.stateRoot);
    if (message.prevStateRoot.length !== 0) writer.uint32(42).bytes(message.prevStateRoot);
    if (message.txCount !== 0n) writer.uint32(48).uint64(message.txCount);
    if (message.dataHash.length !== 0) writer.uint32(58).bytes(message.dataHash);
    if (message.proof.length !== 0) writer.uint32(66).bytes(message.proof);
    if (message.withdrawalsRoot.length !== 0) writer.uint32(74).bytes(message.withdrawalsRoot);
    return writer;
  },
  decode(input: BinaryReader | Uint8Array, length?: number): MsgSubmitBatch {
    const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
    const end = length === undefined ? reader.len : reader.pos + length;
    const message: MsgSubmitBatch = {
      sequencer: "",
      rollupId: "",
      batchIndex: 0n,
      stateRoot: new Uint8Array(),
      prevStateRoot: new Uint8Array(),
      txCount: 0n,
      dataHash: new Uint8Array(),
      proof: new Uint8Array(),
      withdrawalsRoot: new Uint8Array(),
    };
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.sequencer = reader.string();
          break;
        case 2:
          message.rollupId = reader.string();
          break;
        case 3:
          message.batchIndex = reader.uint64();
          break;
        case 4:
          message.stateRoot = reader.bytes();
          break;
        case 5:
          message.prevStateRoot = reader.bytes();
          break;
        case 6:
          message.txCount = reader.uint64();
          break;
        case 7:
          message.dataHash = reader.bytes();
          break;
        case 8:
          message.proof = reader.bytes();
          break;
        case 9:
          message.withdrawalsRoot = reader.bytes();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object: Partial<MsgSubmitBatch>): MsgSubmitBatch {
    return {
      sequencer: object.sequencer ?? "",
      rollupId: object.rollupId ?? "",
      batchIndex: object.batchIndex ?? 0n,
      stateRoot: object.stateRoot ?? new Uint8Array(),
      prevStateRoot: object.prevStateRoot ?? new Uint8Array(),
      txCount: object.txCount ?? 0n,
      dataHash: object.dataHash ?? new Uint8Array(),
      proof: object.proof ?? new Uint8Array(),
      withdrawalsRoot: object.withdrawalsRoot ?? new Uint8Array(),
    };
  },
};

export const MsgChallengeBatch = {
  typeUrl: `${TYPE_URL_PREFIX}MsgChallengeBatch`,
  encode(message: MsgChallengeBatch, writer: BinaryWriter = BinaryWriter.create()): BinaryWriter {
    if (message.challenger !== "") writer.uint32(10).string(message.challenger);
    if (message.rollupId !== "") writer.uint32(18).string(message.rollupId);
    if (message.batchIndex !== 0n) writer.uint32(24).uint64(message.batchIndex);
    if (message.proof.length !== 0) writer.uint32(34).bytes(message.proof);
    return writer;
  },
  decode(input: BinaryReader | Uint8Array, length?: number): MsgChallengeBatch {
    const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
    const end = length === undefined ? reader.len : reader.pos + length;
    const message: MsgChallengeBatch = {
      challenger: "",
      rollupId: "",
      batchIndex: 0n,
      proof: new Uint8Array(),
    };
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.challenger = reader.string();
          break;
        case 2:
          message.rollupId = reader.string();
          break;
        case 3:
          message.batchIndex = reader.uint64();
          break;
        case 4:
          message.proof = reader.bytes();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object: Partial<MsgChallengeBatch>): MsgChallengeBatch {
    return {
      challenger: object.challenger ?? "",
      rollupId: object.rollupId ?? "",
      batchIndex: object.batchIndex ?? 0n,
      proof: object.proof ?? new Uint8Array(),
    };
  },
};

export const MsgResolveChallenge = {
  typeUrl: `${TYPE_URL_PREFIX}MsgResolveChallenge`,
  encode(message: MsgResolveChallenge, writer: BinaryWriter = BinaryWriter.create()): BinaryWriter {
    if (message.resolver !== "") writer.uint32(10).string(message.resolver);
    if (message.rollupId !== "") writer.uint32(18).string(message.rollupId);
    if (message.batchIndex !== 0n) writer.uint32(24).uint64(message.batchIndex);
    if (message.fraudUpheld === true) writer.uint32(32).bool(message.fraudUpheld);
    return writer;
  },
  decode(input: BinaryReader | Uint8Array, length?: number): MsgResolveChallenge {
    const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
    const end = length === undefined ? reader.len : reader.pos + length;
    const message: MsgResolveChallenge = {
      resolver: "",
      rollupId: "",
      batchIndex: 0n,
      fraudUpheld: false,
    };
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.resolver = reader.string();
          break;
        case 2:
          message.rollupId = reader.string();
          break;
        case 3:
          message.batchIndex = reader.uint64();
          break;
        case 4:
          message.fraudUpheld = reader.bool();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object: Partial<MsgResolveChallenge>): MsgResolveChallenge {
    return {
      resolver: object.resolver ?? "",
      rollupId: object.rollupId ?? "",
      batchIndex: object.batchIndex ?? 0n,
      fraudUpheld: object.fraudUpheld ?? false,
    };
  },
};

export const MsgPauseRollup = {
  typeUrl: `${TYPE_URL_PREFIX}MsgPauseRollup`,
  encode(message: MsgPauseRollup, writer: BinaryWriter = BinaryWriter.create()): BinaryWriter {
    if (message.creator !== "") writer.uint32(10).string(message.creator);
    if (message.rollupId !== "") writer.uint32(18).string(message.rollupId);
    if (message.reason !== "") writer.uint32(26).string(message.reason);
    return writer;
  },
  decode(input: BinaryReader | Uint8Array, length?: number): MsgPauseRollup {
    const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
    const end = length === undefined ? reader.len : reader.pos + length;
    const message: MsgPauseRollup = { creator: "", rollupId: "", reason: "" };
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.creator = reader.string();
          break;
        case 2:
          message.rollupId = reader.string();
          break;
        case 3:
          message.reason = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object: Partial<MsgPauseRollup>): MsgPauseRollup {
    return {
      creator: object.creator ?? "",
      rollupId: object.rollupId ?? "",
      reason: object.reason ?? "",
    };
  },
};

export const MsgResumeRollup = {
  typeUrl: `${TYPE_URL_PREFIX}MsgResumeRollup`,
  encode(message: MsgResumeRollup, writer: BinaryWriter = BinaryWriter.create()): BinaryWriter {
    if (message.creator !== "") writer.uint32(10).string(message.creator);
    if (message.rollupId !== "") writer.uint32(18).string(message.rollupId);
    return writer;
  },
  decode(input: BinaryReader | Uint8Array, length?: number): MsgResumeRollup {
    const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
    const end = length === undefined ? reader.len : reader.pos + length;
    const message: MsgResumeRollup = { creator: "", rollupId: "" };
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.creator = reader.string();
          break;
        case 2:
          message.rollupId = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object: Partial<MsgResumeRollup>): MsgResumeRollup {
    return { creator: object.creator ?? "", rollupId: object.rollupId ?? "" };
  },
};

export const MsgStopRollup = {
  typeUrl: `${TYPE_URL_PREFIX}MsgStopRollup`,
  encode(message: MsgStopRollup, writer: BinaryWriter = BinaryWriter.create()): BinaryWriter {
    if (message.creator !== "") writer.uint32(10).string(message.creator);
    if (message.rollupId !== "") writer.uint32(18).string(message.rollupId);
    return writer;
  },
  decode(input: BinaryReader | Uint8Array, length?: number): MsgStopRollup {
    const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
    const end = length === undefined ? reader.len : reader.pos + length;
    const message: MsgStopRollup = { creator: "", rollupId: "" };
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.creator = reader.string();
          break;
        case 2:
          message.rollupId = reader.string();
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object: Partial<MsgStopRollup>): MsgStopRollup {
    return { creator: object.creator ?? "", rollupId: object.rollupId ?? "" };
  },
};

export const MsgExecuteWithdrawal = {
  typeUrl: `${TYPE_URL_PREFIX}MsgExecuteWithdrawal`,
  encode(
    message: MsgExecuteWithdrawal,
    writer: BinaryWriter = BinaryWriter.create(),
  ): BinaryWriter {
    if (message.submitter !== "") writer.uint32(10).string(message.submitter);
    if (message.rollupId !== "") writer.uint32(18).string(message.rollupId);
    if (message.batchIndex !== 0n) writer.uint32(24).uint64(message.batchIndex);
    if (message.withdrawalIndex !== 0n) writer.uint32(32).uint64(message.withdrawalIndex);
    if (message.recipient !== "") writer.uint32(42).string(message.recipient);
    if (message.denom !== "") writer.uint32(50).string(message.denom);
    if (message.amount !== 0n) writer.uint32(56).int64(message.amount);
    for (const v of message.proof) {
      writer.uint32(66).bytes(v);
    }
    return writer;
  },
  decode(input: BinaryReader | Uint8Array, length?: number): MsgExecuteWithdrawal {
    const reader = input instanceof BinaryReader ? input : new BinaryReader(input);
    const end = length === undefined ? reader.len : reader.pos + length;
    const message: MsgExecuteWithdrawal = {
      submitter: "",
      rollupId: "",
      batchIndex: 0n,
      withdrawalIndex: 0n,
      recipient: "",
      denom: "",
      amount: 0n,
      proof: [],
    };
    while (reader.pos < end) {
      const tag = reader.uint32();
      switch (tag >>> 3) {
        case 1:
          message.submitter = reader.string();
          break;
        case 2:
          message.rollupId = reader.string();
          break;
        case 3:
          message.batchIndex = reader.uint64();
          break;
        case 4:
          message.withdrawalIndex = reader.uint64();
          break;
        case 5:
          message.recipient = reader.string();
          break;
        case 6:
          message.denom = reader.string();
          break;
        case 7:
          message.amount = reader.int64();
          break;
        case 8:
          message.proof.push(reader.bytes());
          break;
        default:
          reader.skipType(tag & 7);
          break;
      }
    }
    return message;
  },
  fromPartial(object: Partial<MsgExecuteWithdrawal>): MsgExecuteWithdrawal {
    return {
      submitter: object.submitter ?? "",
      rollupId: object.rollupId ?? "",
      batchIndex: object.batchIndex ?? 0n,
      withdrawalIndex: object.withdrawalIndex ?? 0n,
      recipient: object.recipient ?? "",
      denom: object.denom ?? "",
      amount: object.amount ?? 0n,
      proof: object.proof ?? [],
    };
  },
};

File: packages/rust/src/tx/codecs.rs
//! Protobuf codecs for the `qorechain.rdk.v1` transaction messages, derived with
//! [`prost`].
//!
//! These mirror the on-chain `rdk` module's `tx.proto` exactly -- field numbers,
//! wire types, and message names. 64-bit integers use `i64`/`u64` matching the
//! proto (`int64` for amounts, `uint64` for indices/counts); `bytes` fields use
//! `Vec<u8>`. The type URLs follow the Cosmos `/package.Msg` convention.

use prost::Message;

/// The type-URL prefix for `rdk` messages.
pub const TYPE_URL_PREFIX: &str = "/qorechain.rdk.v1.";

/// `MsgCreateRollup` -- register a new rollup.
#[derive(Clone, PartialEq, Eq, Message)]
pub struct MsgCreateRollup {
    /// Creator (signer) address.
    #[prost(string, tag = "1")]
    pub creator: String,
    /// Unique rollup id.
    #[prost(string, tag = "2")]
    pub rollup_id: String,
    /// Preset profile.
    #[prost(string, tag = "3")]
    pub profile: String,
    /// Execution environment.
    #[prost(string, tag = "4")]
    pub vm_type: String,
    /// Stake committed at creation, in uqor.
    #[prost(int64, tag = "5")]
    pub stake_amount: i64,
}

/// `MsgSubmitBatch` -- submit a settlement batch.
#[derive(Clone, PartialEq, Eq, Message)]
pub struct MsgSubmitBatch {
    /// Sequencer (signer) address.
    #[prost(string, tag = "1")]
    pub sequencer: String,
    /// Target rollup id.
    #[prost(string, tag = "2")]
    pub rollup_id: String,
    /// Batch index.
    #[prost(uint64, tag = "3")]
    pub batch_index: u64,
    /// Post-state root.
    #[prost(bytes = "vec", tag = "4")]
    pub state_root: Vec<u8>,
    /// Previous-state root.
    #[prost(bytes = "vec", tag = "5")]
    pub prev_state_root: Vec<u8>,
    /// Transaction count in the batch.
    #[prost(uint64, tag = "6")]
    pub tx_count: u64,
    /// Data-availability hash.
    #[prost(bytes = "vec", tag = "7")]
    pub data_hash: Vec<u8>,
    /// Settlement proof.
    #[prost(bytes = "vec", tag = "8")]
    pub proof: Vec<u8>,
    /// Binary Merkle root of the batch's L2->L1 messages.
    #[prost(bytes = "vec", tag = "9")]
    pub withdrawals_root: Vec<u8>,
}

/// `MsgChallengeBatch` -- challenge an optimistic batch.
#[derive(Clone, PartialEq, Eq, Message)]
pub struct MsgChallengeBatch {
    /// Challenger (signer) address.
    #[prost(string, tag = "1")]
    pub challenger: String,
    /// Target rollup id.
    #[prost(string, tag = "2")]
    pub rollup_id: String,
    /// Batch index.
    #[prost(uint64, tag = "3")]
    pub batch_index: u64,
    /// Fraud proof.
    #[prost(bytes = "vec", tag = "4")]
    pub proof: Vec<u8>,
}

/// `MsgResolveChallenge` -- resolve an open challenge.
#[derive(Clone, PartialEq, Eq, Message)]
pub struct MsgResolveChallenge {
    /// Resolver (signer) address.
    #[prost(string, tag = "1")]
    pub resolver: String,
    /// Target rollup id.
    #[prost(string, tag = "2")]
    pub rollup_id: String,
    /// Batch index.
    #[prost(uint64, tag = "3")]
    pub batch_index: u64,
    /// Whether fraud was upheld.
    #[prost(bool, tag = "4")]
    pub fraud_upheld: bool,
}

/// `MsgPauseRollup` -- pause an active rollup.
#[derive(Clone, PartialEq, Eq, Message)]
pub struct MsgPauseRollup {
    /// Creator (signer) address.
    #[prost(string, tag = "1")]
    pub creator: String,
    /// Target rollup id.
    #[prost(string, tag = "2")]
    pub rollup_id: String,
    /// Optional reason.
    #[prost(string, tag = "3")]
    pub reason: String,
}

/// `MsgResumeRollup` -- resume a paused rollup.
#[derive(Clone, PartialEq, Eq, Message)]
pub struct MsgResumeRollup {
    /// Creator (signer) address.
    #[prost(string, tag = "1")]
    pub creator: String,
    /// Target rollup id.
    #[prost(string, tag = "2")]
    pub rollup_id: String,
}

/// `MsgStopRollup` -- stop a rollup permanently.
#[derive(Clone, PartialEq, Eq, Message)]
pub struct MsgStopRollup {
    /// Creator (signer) address.
    #[prost(string, tag = "1")]
    pub creator: String,
    /// Target rollup id.
    #[prost(string, tag = "2")]
    pub rollup_id: String,
}

/// `MsgExecuteWithdrawal` -- finalize an L2->L1 withdrawal.
#[derive(Clone, PartialEq, Eq, Message)]
pub struct MsgExecuteWithdrawal {
    /// Submitter (signer) address.
    #[prost(string, tag = "1")]
    pub submitter: String,
    /// Target rollup id.
    #[prost(string, tag = "2")]
    pub rollup_id: String,
    /// Finalized batch index.
    #[prost(uint64, tag = "3")]
    pub batch_index: u64,
    /// Withdrawal index within the batch.
    #[prost(uint64, tag = "4")]
    pub withdrawal_index: u64,
    /// The committed recipient.
    #[prost(string, tag = "5")]
    pub recipient: String,
    /// The withdrawal denom.
    #[prost(string, tag = "6")]
    pub denom: String,
    /// The withdrawal amount.
    #[prost(int64, tag = "7")]
    pub amount: i64,
    /// Sibling hashes from the leaf to `withdrawals_root`.
    #[prost(bytes = "vec", repeated, tag = "8")]
    pub proof: Vec<Vec<u8>>,
}

/// A trait giving each `rdk` message its Cosmos type URL and proto encoding.
pub trait RdkMsg: Message + Sized {
    /// The message's short name (e.g. `MsgCreateRollup`).
    const NAME: &'static str;

    /// The full Cosmos type URL (`/qorechain.rdk.v1.<Name>`).
    fn type_url() -> String {
        format!("{TYPE_URL_PREFIX}{}", Self::NAME)
    }

    /// Encode the message to protobuf bytes.
    fn encode_to_vec_msg(&self) -> Vec<u8> {
        self.encode_to_vec()
    }
}

macro_rules! impl_rdk_msg {
    ($t:ty, $name:literal) => {
        impl RdkMsg for $t {
            const NAME: &'static str = $name;
        }
    };
}

impl_rdk_msg!(MsgCreateRollup, "MsgCreateRollup");
impl_rdk_msg!(MsgSubmitBatch, "MsgSubmitBatch");
impl_rdk_msg!(MsgChallengeBatch, "MsgChallengeBatch");
impl_rdk_msg!(MsgResolveChallenge, "MsgResolveChallenge");
impl_rdk_msg!(MsgPauseRollup, "MsgPauseRollup");
impl_rdk_msg!(MsgResumeRollup, "MsgResumeRollup");
impl_rdk_msg!(MsgStopRollup, "MsgStopRollup");
impl_rdk_msg!(MsgExecuteWithdrawal, "MsgExecuteWithdrawal");
